### ISSUE TITLE
[jit] `jit.annotate` and multiple compilation of methods with different types

### ISSUE THREAD

**Issue by narfanar** [pytorch/pytorch#38744]
## 🚀 Feature
Provide a way to `jit.script` a method more than once with different signatures.

## Motivation
Providing "templating" or specializing generic functionality by reusing function body definitions in Python for several scripted functions that differ only the signature.

Currently the only way seems to write the entire function multiple times with the correct signature embedded in native Python syntax. `jit.annotate` and `jit.Attribute` are also getting deprecated (they never provided this functionality, but the API design is relevant).

## Pitch
Provide a function similar to `jit.script` with an additional argument being a form of type signature for the method being compiled.

This type signature will probably need to be actual Python objects (e.g. based on `Callable[]`) to allow forwarding the types to the compiler (which can't see the context here?).

Example:
```py
foo_1 = jit.script_typed(foo, Callable[[Tensor, Tensor], Tensor]])
foo_2 = jit.script_typed(foo, Callable[[Tensor, Tuple[Tensor, Tensor]], Tensor])
```

### Questions
How do we handle functions taking `self`, `cls`?

## Alternatives
Not sure.

## Additional context
Julia provides multiple dispatch. At a minimum, this provides some generics capability to the language.

In PyTorch, this is already handled by the base type (Tensor) at the element level. But we're missing it for more complex compositions of other available types.

**Comment by narfanar** [comment]
A more commonly relevant example: typing scripted versions of RNN cells whose states change shape based on initialization arguments.

E.g. an LSTM cell with a state that is either:
1. `Tuple[Tensor, Tensor]` (traditional) or
1. `Tuple[Tensor, Tensor, Tensor]` when recurrent dropout is enabled (one possible implementation).

**Comment by janeyx99** [comment]
TorchScript has been deprecated in favor of torch.export as of 2.11. https://pytorch.org/blog/pytorch-2-11-release-blog/ Closing this issue as we are not further developing on it--please ping if you believe this issue was erroneously closed.
